File: lecture17/exercise17-D/template.py
#!/usr/bin/env python3

# Exercise 17-D: Invert Binary Tree

import collections
import sys

# Classes

class Node:
    def __init__(self, value=0, left=None, right=None):
        self.value = value
        self.left  = left
        self.right = right

# Functions

def read_tree(values, index=0):
    # TODO: Use divide and conquer to parse tree
    pass

def walk_tree(root):
    # TODO: Use BFS Traversal with Queue to create list of values
    pass

def dump_tree(root):
    print(','.join(map(str, walk_tree(root))))

def invert_tree(root):
    # TODO: Use divide and conquer to invert binary tree
    pass

# Main Execution

def main():
    for line in sys.stdin:
        values = list(map(int, line.split()))
        root   = read_tree(values)

        invert_tree(root)
        dump_tree(root)

if __name__ == '__main__':
    main()

File: lecture17/exercise17-D/template.cpp
// Exercise 17-D: Invert Binary Tree

#include <iostream>
#include <memory>
#include <queue>
#include <string>
#include <sstream>
#include <vector>

using namespace std;

// Node structures

template <typename T>
struct Node {
    T value;
    struct Node *left;
    struct Node *right;
};

typedef Node<int> IntNode;

// Functions

bool read_values(vector<int> &values) {
    values.clear();

    string line;
    if (getline(cin, line)) {
    	stringstream ss(line);
    	int i;

    	while (ss >> i) {
    	    values.push_back(i);
	}
    }

    return !values.empty();
}

IntNode*  read_tree(vector<int> &values, size_t index=0) {
    // TODO: Use divide and conquer to parse tree
    return NULL
}

void	dump_tree(IntNode *root) {
    // TODO: Use BFS traversal with queue to dump values in tree
}

void	invert_tree(IntNode *root) {
    // TODO: Use divide and conquer to invert binary tree
}

// Main execution

int main(int argc, char *argv[]) {
    vector<int> values;

    while (read_values(values)) {
    	auto root = read_tree(values);
    	invert_tree(root);
    	dump_tree(root);
    }

    return 0;
}
